### PR TITLE
Proof of Concept on Bugprone Narrowing Conversions: Base64

### PR DESCRIPTION
**Pull request by devbali**
This is a small commit that is an example of the majority of changes required for clangtidy's Bugprone Narrowing Conversions. I will open more small Pull Requests periodically if this is the appropriate approach.

### CODE REVIEW

**Comment by 0xxon** [comment]
Out of curiosity - in this patch, all the changes seem to be lengths/sizes. Is there a reason to prefer uint64_t to e.g. size_t - which can be a different size on different platform, but always is guaranteed to be able to express the maximum size of any object.

**Comment by devbali** [comment]
Yeah, we could totally use size_t as well. Should I use it instead of uint64_t?

**Comment by devbali** [comment]
I think our decision to use uint64_t over size_t was mostly arbitrary but @timwoj correct me if I'm wrong.

**Comment by timwoj** [comment]
If it's related to the size of a container, definitely use whatever the container defines it as. I'm going to get a full review of it today.

**Comment by timwoj** [comment]
> Out of curiosity - in this patch, all the changes seem to be lengths/sizes. Is there a reason to prefer uint64_t to e.g. size_t - which can be a different size on different platform, but always is guaranteed to be able to express the maximum size of any object.

It was pointer math that triggered the warning fixed in this patch at https://github.com/zeek/zeek/blob/master/src/Base64.cc#L196. Ideally we'd use `ptrdiff_t` here for those kind of changes but iirc we discussed that and decided against using it. `size_t` is a reasonable substitute since it should be the same size everywhere.

File: src/analyzer/protocol/mime/MIME.cc [hunk 1]
```diff
@@ -1131,7 +1131,7 @@ void MIME_Entity::DecodeBase64(int len, const char* data)
 		{
 		rlen = 128;
 		char* prbuf = rbuf;
-		int decoded = base64_decoder->Decode(len, data, &rlen, &prbuf);
+		int decoded = base64_decoder->Decode(len, data, reinterpret_cast<uint64_t*>(&rlen), &prbuf); // typecasting for now
```
Comment (0xxon) [inline src/analyzer/protocol/mime/MIME.cc]: This does not look right to me.

This takes a pointer to an int (length can be 2 or 4 bytes, depending on system), and casts it to an uint64_t (which has a length of 4 bytes). In cases where sizeof(int) != sizeof(uint64_t), that reads into the following memory space, right?

Also - this should probably be size_t too now (which does not fix this problem).

**Comment by 0xxon** [comment]
Added a comment - I think this now results in a buffer overread in certain circumstances.

**Comment by devbali** [comment]
I was using this for unsigned int first, then changed it to uint64_t without thinking too much - something I should have done. So sorry about that. Yikes.

P.S. Also sorry for the bad commit message double yikes

**Comment by 0xxon** [comment]
This now complains about a comparison in Base64.cc:

```
../src/Base64.cc: In function ‘BroString* decode_base64(const BroString*, const BroString*, Connection*)’:
../src/Base64.cc:233:68: warning: comparison of integer expressions of different signedness: ‘size_t’ {aka ‘long unsigned int’} and ‘int’ [-Wsign-compare]
  233 |  if ( dec.Decode(s->Len(), (const char*) s->Bytes(), &rlen, &rbuf) == -1 )
      |       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~
```

And it seems to be right. However, this code never seems to have worked - Decode does not seem to return a -1 in an error case - even before this change. It just logs to reporter.

I am a bit torn on what to do here. We can either just get rid of the check and have things work as before - it just returns a partially decoded string. Or we can actually have Decode return an error case. decode_base64 is currently written in a way that would return an empty string in this case (instead of a partial decode).

Does anyone have an opinion before I select one of the two?

File: src/analyzer/protocol/mime/MIME.cc [hunk 2]
```diff
@@ -1131,7 +1131,8 @@ void MIME_Entity::DecodeBase64(int len, const char* data)
 		{
 		rlen = 128;
 		char* prbuf = rbuf;
-		int decoded = base64_decoder->Decode(len, data, &rlen, &prbuf);
+		size_t x = static_cast<size_t> (rlen);
+		int decoded = base64_decoder->Decode(len, data, &x, &prbuf); // typecasting for now
```
Comment (0xxon) [inline src/analyzer/protocol/mime/MIME.cc]: I am fixing this in the merge. But note - this leads to wrong results. x is changed when it is passed to decode - and your re-assignment shadows this and does not reflect the change in rlen.

**Pull request merged**